## -----------------------------------------------------------------------------
## Purpose of script: Extract HLA haplotypes from relevant patients and format
##                    for netMHCpan
##
## Author: Oliver Artz
## Date Created:
## Email: artzo@mskcc.org
##
## -----------------------------------------------------------------------------

# clear work space -------------------------------------------------------------
# rm(list = ls())
# gc()

# load libraries ---------------------------------------------------------------
options(repos = "https://cran.rstudio.com")
if (!require('pacman')) install.packages('pacman', dependencies = TRUE); library(pacman)

p_load(tidyverse, data.table, here)

# set parameters ---------------------------------------------------------------
project_root <- here()
project_folder <- "20250521_TMBH_acquisition"
project_dir <- paste0(project_root, "/../", project_folder)

# directory that has all the HPC processed data
output_dir <- paste0(project_dir, "/data/processed/hpc/")


# directory for fasta files generated by SNPEff
fasta_dir <- paste0(output_dir, "/fasta/patient_status_compared_to_baseline")

# load data --------------------------------------------------------------------
# all haplotypes
all_haplotypes <- fread("/data/ldiaz/artzo/assets/IMPACT/HLA/impact_hla_manifest_2025_01_16.txt")

# patient key
patient_key_mod <- fread(paste0(project_dir, "/data/processed/001_patient_key_mod.txt"))

# wrangle ----------------------------------------------------------------------
# filter haplotype data for relevant patients
hla_haplotypes_mod <- data.frame(patient = patient_key_mod$dmp_patient_id) %>% 
  left_join(all_haplotypes,
            by = c("patient")) %>% 
  distinct(patient, .keep_all = TRUE)

# make long format
haplotype_info_long <- hla_haplotypes_mod %>% 
  pivot_longer(cols = c("A1", "A2", "B1", "B2", "C1", "C2"), names_to = "gene", values_to = "haplotype") %>% 
  filter

# translate into HLA alleles for netMHCpan
translate_haplotype <- function(tempo_haplotype){
  tempo_haplotype <- str_split_1(tempo_haplotype, "_")
  netmhc_haplotype <- paste0("HLA-", toupper(tempo_haplotype[2]), tempo_haplotype[3], ":", tempo_haplotype[4])
  return(netmhc_haplotype)
}

# translate all haplotypes
haplotype_netmhc <- haplotype_info_long %>% 
  mutate(haplotype = map(haplotype, translate_haplotype)) %>% 
  unnest(cols = haplotype)

# add patient ID
idx <- match(haplotype_netmhc$patient, patient_key_mod$dmp_patient_id)
haplotype_netmhc$patient <- as.numeric(patient_key_mod$patient_id[idx])

# get FASTA file info
fasta_files <- list.files(path = fasta_dir, 
                          pattern = ".fasta") %>% 
  str_remove(".fasta")

# make final data frame for each FASTA sample we have
df_final <- data.frame(sample = fasta_files) %>% 
  separate(sample, into = c("patient", "status")) %>% 
  mutate(patient = as.numeric(patient)) %>% 
  left_join(haplotype_netmhc,
            by = c("patient" = "patient")) %>% 
  mutate(sample = paste0(patient, "_", status)) %>% 
  dplyr::select(sample, haplotype)

# if a patient is homozygous, we only want to predict one allele. if we predict two
# alleles in parallel, netMHCpan will append the output files which can lead to 
# issues in downstream analyses
df_final <- df_final %>%
  group_by(sample, haplotype) %>%
  slice(1) %>%
  ungroup()

# remove patients without HLA info
df_final <- df_final %>% 
  filter(!is.na(haplotype))

# QC ---------------------------------------------------------------------------
# check patients with HLA haplotype data in final dataset
patients_with_hla <- df_final %>% 
  distinct(sample) %>% 
  separate(sample, into = c("patient", "status"), sep = "_", remove = FALSE) %>% 
  mutate(patient = as.numeric(patient)) %>% 
  arrange(patient)

cat("Number of samples with HLA data in df_final:", nrow(patients_with_hla), "\n")
cat("Number of unique patients with HLA data in df_final:", length(unique(patients_with_hla$patient)), "\n")
cat("Patients with HLA data in df_final:", paste(sort(unique(patients_with_hla$patient)), collapse = ", "), "\n")

# Check which patients from FASTA files do NOT have HLA info
all_fasta_patients <- data.frame(sample = fasta_files) %>% 
  separate(sample, into = c("patient", "status")) %>% 
  mutate(patient = as.numeric(patient)) %>% 
  distinct(patient) %>% 
  arrange(patient)

patients_without_hla <- setdiff(all_fasta_patients$patient, unique(patients_with_hla$patient))

if(length(patients_without_hla) > 0) {
  cat("\nPatients with FASTA files but NO HLA data:", paste(sort(patients_without_hla), collapse = ", "), "\n")
} else {
  cat("\nAll patients with FASTA files have HLA data available\n")
}

cat("\nTotal patients with FASTA files:", nrow(all_fasta_patients), "\n")
cat("Total patients with both FASTA files AND HLA data:", length(unique(patients_with_hla$patient)), "\n")

# export -----------------------------------------------------------------------
write.table(
  df_final,
  file = paste0(output_dir, "hla_haplotypes.txt"),
  quote = FALSE,
  sep = "\t",
  col.names = FALSE,
  row.names = FALSE
)

# cleanup ----------------------------------------------------------------------
rm()